Fix race condition in system.query_metric_log

pamarcos opened the pull request:
If a scheduled task was pending to lock the query_mutex, it could do so between the lock.unlock() and lock.lock() of finishQuery which is used to break the exec_mutex deadlock.

In that case, a task scheduled running after the task finished could attempt to do a new collection, which is wrong. Let's avoid that altogether, and also add some defensive programming to ensure this does not happen in the future.

Closes #71244

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix wrong value in system.query_metric_log due to unexpected race condition

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

pamarcos committed:
Fix race condition in system.query_metric_log

If a scheduled task was pending to lock the query_mutex,
it could do so between the lock.unlock() and lock.lock()
of finishQuery which is used to break the exec_mutex deadlock.

In that case, a task scheduled running after the task
finished could attempt to do a new collection, which
is wrong. Let's avoid that altogether, and also add
some defensive programming to ensure this does not happen
in the future.

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #71124 from ClickHouse/ClickHouse:hotfix_query_metric_log -->
*This is an automated comment for commit de880d0f8ba39258f4e6080cfb8927d2243dcce9 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71124/de880d0f8ba39258f4e6080cfb8927d2243dcce9/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/c82cf25b3e5864bcc153cbe45adb8c6527e1ec6e/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

pamarcos committed:
Remove wrong LOGICAL_ERROR

pamarcos committed:
Fix another source of race condition

At first I thought of doing the bare minimum when
`queries_mutex` was locked. However, scheduling
the task does not lock `exec_mutex` but other ones.

So, there is no deadlock in sight of scheduling
within `queries_mutex`. This way, we prevent a rare
case in which the task is scheduled before the query
is even added to the queries list.

pamarcos committed:
fixup! Remove wrong LOGICAL_ERROR

pamarcos commented:
I've removed the `LOGICAL_ERROR` I introduced to prevent errors in the future because of this unexpected thing:

```
ch_test 02122_join_group_by_timeout

2024.10.28 11:26:26.265549 [ 818399 ] {vtqspuywkjqw} <Fatal> : Logical error: 'Profile event count is not monotonically increasing for 'QueryTimeMicroseconds': new value 0 is smaller
 than previous value 1000754'.
```

I'm investigating :eyes:

pamarcos committed:
Use the proper timestamp from where the query_info is collected

pamarcos committed:
Fix for collecting old metrics already collected

pamarcos commented:
Ok, I think I found the issue. It happened for those queries where the collecting task was scheduled at a similar time as the query finished. I found it by logging when the same condition as the `LOGICAL_ERROR` happened and inspecting those queries.

```bash
❯ grep -ina "QueryMetricLog" /var/log/clickhouse-server/clickhouse-server.log | grep "srrbmequdcwe"                                                                                   
595:2024.10.28 15:05:40.172239 [ 941982 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query srrbmequdcwe. Schedule next: true                                          
611:2024.10.28 15:05:40.175919 [ 941959 ] {srrbmequdcwe} <Debug> QueryMetricLog: Collecting query_metric_log for query srrbmequdcwe. Schedule next: false                             

❯ grep -ina "QueryMetricLog" /var/log/clickhouse-server/clickhouse-server.log | grep "88c23f18-b9fc-4b38-a073-ad540a9b68f4"                                                           
930:2024.10.28 15:05:42.549816 [ 942051 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 88c23f18-b9fc-4b38-a073-ad540a9b68f4. Schedule next: true                  
952:2024.10.28 15:05:42.557082 [ 941958 ] {88c23f18-b9fc-4b38-a073-ad540a9b68f4} <Debug> QueryMetricLog: Collecting query_metric_log for query 88c23f18-b9fc-4b38-a073-ad540a9b68f4. S
chedule next: false     
```

In those cases, it may happen that a slightly older metrics is flushed. I fixed it by:

1. Making sure the `timestamp` used is always considered when the `QueryStatusInfo` is collected: [`961fa4d` (#71124)](https://github.com/ClickHouse/ClickHouse/pull/71124/commits/961fa4d5ce0d7d7494b358d927e617dc7ed19eb5)
2. Skipping metrics collections of older timestamps: [`66013a2` (#71124)](https://github.com/ClickHouse/ClickHouse/pull/71124/commits/66013a2bb9e8ff567e1e2ac448fdee7cce7465a3)

With those two things, I've added back the beloved LOGICAL_ERROR that helped finding this unexpected behavior

pamarcos committed:
Merge branch 'master' into hotfix_query_metric_log

pamarcos committed:
Remove the query from the list as soon as we can

There's no need to take again the lock. This comes from
a prior version where the logic needed to be different.

pamarcos committed:
Add trace to help debugging the issue on the CI

pamarcos committed:
Add timestamp of QueryInfoStatus to debug trace to ease debugging

pamarcos committed:
Revert "Add trace to help debugging the issue on the CI"

This reverts commit d063fc0a1263ed5f0a799156aafc5ec830155a7e.

pamarcos committed:
Merge branch 'master' into hotfix_query_metric_log

pamarcos commented:
I've found something weird happening on the CI, which is why I've added two new commits to try to clarify why it happens.

A counter that didn't increase monotonically. Potential old value collected after a new one:
```
fd "clickhouse-server*.log" integration_run_parallel1_0 | xargs zstdgrep -C 10 "QueryMetricLog" | grep -i fatal
integration_run_parallel1_0/test_hedged_requests/_instances-0-gw2/node/logs/clickhouse-server.log.4.gz-2024.10.28 20:42:37.898484 [ 8699 ] {} <Fatal> : Logical error: 'Profile event
count is not monotonically increasing for 'NetworkReceiveBytes': new value 290 is smaller than previous value 291'.

fd "clickhouse-server*.log" integration_run_parallel1_0 | xargs zstdgrep "QueryMetricLog" | grep "047368f2-31ae-4a56-aa4f-ace7e76df094"
integration_run_parallel1_0/test_hedged_requests/_instances-0-gw2/node/logs/clickhouse-server.log.4.gz:2024.10.28 20:42:34.907262 [ 8654 ] {} <Debug> QueryMetricLog: Collecting query
_metric_log for query 047368f2-31ae-4a56-aa4f-ace7e76df094. Schedule next: true
integration_run_parallel1_0/test_hedged_requests/_instances-0-gw2/node/logs/clickhouse-server.log.4.gz:2024.10.28 20:42:35.898543 [ 8669 ] {} <Debug> QueryMetricLog: Collecting query
_metric_log for query 047368f2-31ae-4a56-aa4f-ace7e76df094. Schedule next: true
integration_run_parallel1_0/test_hedged_requests/_instances-0-gw2/node/logs/clickhouse-server.log.4.gz:2024.10.28 20:42:36.898809 [ 8688 ] {} <Debug> QueryMetricLog: Collecting query
_metric_log for query 047368f2-31ae-4a56-aa4f-ace7e76df094. Schedule next: true
integration_run_parallel1_0/test_hedged_requests/_instances-0-gw2/node/logs/clickhouse-server.log.4.gz:2024.10.28 20:42:37.898245 [ 8699 ] {} <Debug> QueryMetricLog: Collecting query
_metric_log for query 047368f2-31ae-4a56-aa4f-ace7e76df094. Schedule next: true
```

A counter that didn't increase monotonically. Potential old value collected after a new one.
There shouldn't be a schedule next: true after a schedule next: false. It's as if the periodic scheduled task was scheduled after the query finished, which is impossible according to the logic :cry: 

```
fd "clickhouse-server*.log" integration_run_parallel3_0 | xargs zstdgrep -C 10 "QueryMetricLog" | grep -i fatal
integration_run_parallel3_0/test_secure_socket/_instances-0-gw3/node1/logs/clickhouse-server.log-2024.10.28 20:58:43.398848 [ 9 ] {9c37dfb6-f090-4876-bf43-9f5e3cea1928} <Fatal> : Log
ical error: 'Profile event count is not monotonically increasing for 'NetworkReceiveBytes': new value 136 is smaller than previous value 137'.

fd "clickhouse-server*.log" integration_run_parallel3_0 | xargs zstdgrep "QueryMetricLog" | grep "9c37dfb6-f090-4876-bf43-9f5e3cea1928"
integration_run_parallel3_0/test_secure_socket/_instances-0-gw3/node1/logs/clickhouse-server.log:2024.10.28 20:58:39.314662 [ 31 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 9c37dfb6-f090-4876-bf43-9f5e3cea1928. Schedule next: true
integration_run_parallel3_0/test_secure_socket/_instances-0-gw3/node1/logs/clickhouse-server.log:2024.10.28 20:58:40.310184 [ 32 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 9c37dfb6-f090-4876-bf43-9f5e3cea1928. Schedule next: true
integration_run_parallel3_0/test_secure_socket/_instances-0-gw3/node1/logs/clickhouse-server.log:2024.10.28 20:58:41.309773 [ 33 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 9c37dfb6-f090-4876-bf43-9f5e3cea1928. Schedule next: true
integration_run_parallel3_0/test_secure_socket/_instances-0-gw3/node1/logs/clickhouse-server.log:2024.10.28 20:58:42.309296 [ 35 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 9c37dfb6-f090-4876-bf43-9f5e3cea1928. Schedule next: true
integration_run_parallel3_0/test_secure_socket/_instances-0-gw3/node1/logs/clickhouse-server.log:2024.10.28 20:58:43.309955 [ 38 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 9c37dfb6-f090-4876-bf43-9f5e3cea1928. Schedule next: true
integration_run_parallel3_0/test_secure_socket/_instances-0-gw3/node1/logs/clickhouse-server.log:2024.10.28 20:58:43.398694 [ 9 ] {9c37dfb6-f090-4876-bf43-9f5e3cea1928} <Debug> QueryMetricLog: Collecting query_metric_log for query 9c37dfb6-f090-4876-bf43-9f5e3cea1928. Schedule next: false
integration_run_parallel3_0/test_secure_socket/_instances-0-gw3/node1/logs/clickhouse-server.log:2024.10.28 20:58:44.309403 [ 55 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 9c37dfb6-f090-4876-bf43-9f5e3cea1928. Schedule next: true
```

A counter that didn't increase monotonically. Potential old value collected after a new one.
```
fd "clickhouse-server*.log" integration_run_parallel1_0_2 | xargs zstdgrep -C 10 "QueryMetricLog" | grep -i fatal -C 5
integration_run_parallel1_0_2/test_hedged_requests/_instances-0-gw2/node/logs/clickhouse-server.log.4.gz:2024.10.28 22:07:13.851886 [ 8655 ] {} <Debug> QueryMetricLog: Collecting que
ry_metric_log for query fe370d3d-2c47-44b3-a058-d80f5034d9c3. Schedule next: true
integration_run_parallel1_0_2/test_hedged_requests/_instances-0-gw2/node/logs/clickhouse-server.log.4.gz-2024.10.28 22:07:13.852196 [ 8655 ] {} <Fatal> : Logical error: 'Profile even
t count is not monotonically increasing for 'NetworkReceiveBytes': new value 290 is smaller than previous value 291'.

fd "clickhouse-server*.log" integration_run_parallel1_0_2 | xargs zstdgrep "QueryMetricLog" | grep "fe370d3d-2c47-44b3-a058-d80f5034d9c3"

integration_run_parallel1_0_2/test_hedged_requests/_instances-0-gw2/node/logs/clickhouse-server.log.4.gz:2024.10.28 22:07:10.859559 [ 8616 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query fe370d3d-2c47-44b3-a058-d80f5034d9c3. Schedule next: true
integration_run_parallel1_0_2/test_hedged_requests/_instances-0-gw2/node/logs/clickhouse-server.log.4.gz:2024.10.28 22:07:11.850993 [ 8631 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query fe370d3d-2c47-44b3-a058-d80f5034d9c3. Schedule next: true
integration_run_parallel1_0_2/test_hedged_requests/_instances-0-gw2/node/logs/clickhouse-server.log.4.gz:2024.10.28 22:07:12.851436 [ 8645 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query fe370d3d-2c47-44b3-a058-d80f5034d9c3. Schedule next: true
integration_run_parallel1_0_2/test_hedged_requests/_instances-0-gw2/node/logs/clickhouse-server.log.4.gz:2024.10.28 22:07:13.851886 [ 8655 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query fe370d3d-2c47-44b3-a058-d80f5034d9c3. Schedule next: true
```

Of course, removing the `LOGICAL_ERROR` would get rid of all these altogether, but I'm afraid there might be something else I'm missing happening.

pamarcos committed:
Take into account that some profile events such as `NetworkReceiveBytes` are inaccurate

pamarcos commented:
Update: I've found that `NetworkReceiveBytes` can be inaccurate: https://github.com/ClickHouse/ClickHouse/blob/8246672945423065419baa02c5941dc91c67c3bf/src/IO/ReadBufferFromPocoSocket.cpp#L41

Turns out that this is 100% reproducible with `test_hedged_requests/test.py::test_receive_timeout2` and `test_secure_socket::test` integration tests:

```
sudo find tests/integration/test_hedged_requests -iname "clickhouse-server*.log" | sudo xargs grep -i QueryMetricLog | grep -i 48f69e5d-062e-411b-a4f8-0d9df672      
tests/integration/test_hedged_requests/_instances/node/logs/clickhouse-server.log:2024.10.29 15:12:31.217387 [ 803 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 48f69e5d-062e-411b-a4f8-0d9df672b310 with QueryStatusInfo from 2024.10.29 15:12:31.00217. Schedule next: true
tests/integration/test_hedged_requests/_instances/node/logs/clickhouse-server.log:2024.10.29 15:12:32.215588 [ 816 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 48f69e5d-062e-411b-a4f8-0d9df672b310 with QueryStatusInfo from 2024.10.29 15:12:32.00215. Schedule next: true
tests/integration/test_hedged_requests/_instances/node/logs/clickhouse-server.log:2024.10.29 15:12:33.215772 [ 831 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 48f69e5d-062e-411b-a4f8-0d9df672b310 with QueryStatusInfo from 2024.10.29 15:12:33.00215. Schedule next: true
tests/integration/test_hedged_requests/_instances/node/logs/clickhouse-server.log:2024.10.29 15:12:34.215983 [ 843 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 48f69e5d-062e-411b-a4f8-0d9df672b310 with QueryStatusInfo from 2024.10.29 15:12:34.00215. Schedule next: true
tests/integration/test_hedged_requests/_instances/node/logs/clickhouse-server.log:2024.10.29 15:12:34.216043 [ 843 ] {} <Trace> QueryMetricLog: Query 48f69e5d-062e-411b-a4f8-0d9df672b310 has a more recent metrics collected because 'NetworkReceiveBytes' new value 290 is smaller than previous value 291
tests/integration/test_hedged_requests/_instances/node/logs/clickhouse-server.log:2024.10.29 15:12:34.224626 [ 747 ] {48f69e5d-062e-411b-a4f8-0d9df672b310} <Debug> QueryMetricLog: Collecting query_metric_log for query 48f69e5d-062e-411b-a4f8-0d9df672b310 with QueryStatusInfo from 2024.10.29 15:12:34.00224. Schedule next: false     
```

```
sudo find tests/integration/test_secure_socket -iname "clickhouse-server*.log" | sudo xargs grep -i QueryMetricLog | grep -i baa757fd-509c-43cd-b8dd-5e92615d05d6       
tests/integration/test_secure_socket/_instances/node1/logs/clickhouse-server.log:2024.10.29 15:17:13.672982 [ 54 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query baa757fd-509c-43cd-b8dd-5e92615d05d6 with QueryStatusInfo from 2024.10.29 15:17:13.00672. Schedule next: true
tests/integration/test_secure_socket/_instances/node1/logs/clickhouse-server.log:2024.10.29 15:17:14.671236 [ 66 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query baa757fd-509c-43cd-b8dd-5e92615d05d6 with QueryStatusInfo from 2024.10.29 15:17:14.00671. Schedule next: true
tests/integration/test_secure_socket/_instances/node1/logs/clickhouse-server.log:2024.10.29 15:17:15.671521 [ 77 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query baa757fd-509c-43cd-b8dd-5e92615d05d6 with QueryStatusInfo from 2024.10.29 15:17:15.00671. Schedule next: true
tests/integration/test_secure_socket/_instances/node1/logs/clickhouse-server.log:2024.10.29 15:17:16.671760 [ 89 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query baa757fd-509c-43cd-b8dd-5e92615d05d6 with QueryStatusInfo from 2024.10.29 15:17:16.00671. Schedule next: true
tests/integration/test_secure_socket/_instances/node1/logs/clickhouse-server.log:2024.10.29 15:17:17.671931 [ 98 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query baa757fd-509c-43cd-b8dd-5e92615d05d6 with QueryStatusInfo from 2024.10.29 15:17:17.00671. Schedule next: true
tests/integration/test_secure_socket/_instances/node1/logs/clickhouse-server.log:2024.10.29 15:17:17.685783 [ 9 ] {baa757fd-509c-43cd-b8dd-5e92615d05d6} <Debug> QueryMetricLog: Collecting query_metric_log for query baa757fd-509c-43cd-b8dd-5e92615d05d6 with QueryStatusInfo from 2024.10.29 15:17:17.00680. Schedule next: false
tests/integration/test_secure_socket/_instances/node1/logs/clickhouse-server.log:2024.10.29 15:17:17.685792 [ 9 ] {baa757fd-509c-43cd-b8dd-5e92615d05d6} <Trace> QueryMetricLog: Query baa757fd-509c-43cd-b8dd-5e92615d05d6 has a more recent metrics collected because 'NetworkReceiveBytes' new value 136 is smaller than previous value 137
```

This commit takes this into account: [`7a60543` (#71124)](https://github.com/ClickHouse/ClickHouse/pull/71124/commits/7a60543670259ee39ac343dd0339a712d1e245aa)

pamarcos commented:
I've checked the CI logs and everything seems fine. A few interesting cases that I thought were very unlikely but happen.
Commenting them here for reference:

## Time jumps when using same query_id

This is due to the different queries using the same query_id. It scared the sh*t out of me because I thought tasks were not scheduled properly

```sql
SELECT event_time_microseconds
FROM file('query_metric_log.tsv.zst')
WHERE query_id = '02366_10'                                                                
                                                                                           
Query id: d2350561-6845-44de-a254-5f65bba305d6
                                                                                                                                                                                      
    ┌────event_time_microseconds─┐
 1. │ 2024-10-29 13:24:02.417934 │
 2. │ 2024-10-29 13:24:03.179531 │
 3. │ 2024-10-29 13:24:04.179886 │                                                         
 4. │ 2024-10-29 13:24:33.450475 │                                                         
 5. │ 2024-10-29 13:24:34.412108 │
 6. │ 2024-10-29 13:24:35.420673 │                                                                                                                                                    
 7. │ 2024-10-29 13:24:36.417288 │
 8. │ 2024-10-29 13:24:37.413034 │                                                                                                                                                    
 9. │ 2024-10-29 13:24:38.414281 │
10. │ 2024-10-29 13:24:39.412719 │
11. │ 2024-10-29 13:24:40.412287 │                                                                                                                                                    
12. │ 2024-10-29 13:24:41.421552 │
13. │ 2024-10-29 13:24:42.413777 │
14. │ 2024-10-29 13:24:43.412411 │
15. │ 2024-10-29 13:24:44.414022 │
16. │ 2024-10-29 13:24:45.412828 │
17. │ 2024-10-29 13:24:46.415105 │
18. │ 2024-10-29 13:24:47.412276 │
19. │ 2024-10-29 13:24:48.412889 │
20. │ 2024-10-29 13:24:49.418399 │                                                         
21. │ 2024-10-29 13:24:50.414900 │
22. │ 2024-10-29 13:24:50.931234 │
    └────────────────────────────┘                                                         

22 rows in set. Elapsed: 0.208 sec. 

SELECT                      
    event_time_microseconds,
    type,                                    
    query_duration_ms,                                                                                                                                                                
    exception                 
FROM file('query_log.tsv.zst')               
WHERE (query_id = '02366_10') AND (type > 1)
FORMAT Vertical

Query id: 0a7c12cc-606c-4b67-9ad3-bc702386bda4

Row 1:
──────
event_time_microseconds: 2024-10-29 13:24:04.057542
type:                    ExceptionWhileProcessing
query_duration_ms:       2877
exception:               Code: 394. DB::Exception: Query was cancelled. (QUERY_WAS_CANCELLED) (version 24.11.1.197)

Row 2:
──────
event_time_microseconds: 2024-10-29 13:24:50.931336
type:                    ExceptionWhileProcessing
query_duration_ms:       18518
exception:               Code: 394. DB::Exception: Query was cancelled. (QUERY_WAS_CANCELLED) (version 24.11.1.197)

2 rows in set. Elapsed: 3.264 sec. Processed 284.06 thousand rows, 26.07 MB (87.02 thousand rows/s., 7.99 MB/s.)
Peak memory usage: 272.25 MiB.
```

## A collecting task with schedule=true after a the query finished with schedule=false

Checking the timestamps of the logs and the `QueryStatusInfo`, everything's good

```bash
zstdgrep -ina "QueryMetricLog" clickhouse-server.log.zst | grep -i "45a44d32-8a03-4ba3-817f-ab508ff769e5"                                                                           
141188:2024.10.29 13:13:25.903197 [ 8080 ] {45a44d32-8a03-4ba3-817f-ab508ff769e5} <Debug> QueryMetricLog: Collecting query_metric_log for query 45a44d32-8a03-4ba3-817f-ab508ff769e5 w
ith QueryStatusInfo from 2024.10.29 13:13:25.00902. Schedule next: false                                                                                                              
141193:2024.10.29 13:13:25.903613 [ 947 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 45a44d32-8a03-4ba3-817f-ab508ff769e5 with QueryStatusInfo from 2024.10.29 1
3:13:25.00903. Schedule next: true                                                                                                                                                    
141203:2024.10.29 13:13:25.905395 [ 947 ] {} <Trace> QueryMetricLog: Query 45a44d32-8a03-4ba3-817f-ab508ff769e5 finished already while this collecting task was running 
```

## A query finished right after a collected task with more recent data

The timestamp for the `QueryStatusInfo` of the collecting task contains more recent data

```bash
zstdgrep -ina "QueryMetricLog" clickhouse-server.log.zst | grep -i "0a761af5-2618-4751-823d-4903ce972706"
968724:2024.10.29 13:22:12.138637 [ 1001 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 0a761af5-2618-4751-823d-4903ce972706 with QueryStatusInfo from 2024.10.29 13:22:12.00138. Schedule next: true
969413:2024.10.29 13:22:13.075251 [ 980 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 0a761af5-2618-4751-823d-4903ce972706 with QueryStatusInfo from 2024.10.29 13:22:13.00074. Schedule next: true
970023:2024.10.29 13:22:14.075310 [ 937 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 0a761af5-2618-4751-823d-4903ce972706 with QueryStatusInfo from 2024.10.29 13:22:14.00075. Schedule next: true
970815:2024.10.29 13:22:15.074851 [ 889 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 0a761af5-2618-4751-823d-4903ce972706 with QueryStatusInfo from 2024.10.29 13:22:15.00074. Schedule next: true
971592:2024.10.29 13:22:16.075303 [ 885 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 0a761af5-2618-4751-823d-4903ce972706 with QueryStatusInfo from 2024.10.29 13:22:16.00075. Schedule next: true
972445:2024.10.29 13:22:17.074851 [ 858 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 0a761af5-2618-4751-823d-4903ce972706 with QueryStatusInfo from 2024.10.29 13:22:17.00074. Schedule next: true
973327:2024.10.29 13:22:18.077325 [ 816 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 0a761af5-2618-4751-823d-4903ce972706 with QueryStatusInfo from 2024.10.29 13:22:18.00077. Schedule next: true
974045:2024.10.29 13:22:19.075215 [ 1106 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 0a761af5-2618-4751-823d-4903ce972706 with QueryStatusInfo from 2024.10.29 13:22:19.00074. Schedule next: true
974776:2024.10.29 13:22:20.074927 [ 749 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 0a761af5-2618-4751-823d-4903ce972706 with QueryStatusInfo from 2024.10.29 13:22:20.00074. Schedule next: true
975566:2024.10.29 13:22:21.074563 [ 742 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 0a761af5-2618-4751-823d-4903ce972706 with QueryStatusInfo from 2024.10.29 13:22:21.00074. Schedule next: true
976612:2024.10.29 13:22:22.075104 [ 700 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 0a761af5-2618-4751-823d-4903ce972706 with QueryStatusInfo from 2024.10.29 13:22:22.00074. Schedule next: true
977392:2024.10.29 13:22:23.074678 [ 1144 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 0a761af5-2618-4751-823d-4903ce972706 with QueryStatusInfo from 2024.10.29 13:22:23.00074. Schedule next: true
978285:2024.10.29 13:22:24.075286 [ 1109 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 0a761af5-2618-4751-823d-4903ce972706 with QueryStatusInfo from 2024.10.29 13:22:24.00075. Schedule next: true
979002:2024.10.29 13:22:25.074964 [ 1064 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query 0a761af5-2618-4751-823d-4903ce972706 with QueryStatusInfo from 2024.10.29 13:22:25.00074. Schedule next: true
979005:2024.10.29 13:22:25.077025 [ 8102 ] {0a761af5-2618-4751-823d-4903ce972706} <Debug> QueryMetricLog: Collecting query_metric_log for query 0a761af5-2618-4751-823d-4903ce972706 with QueryStatusInfo from 2024.10.29 13:22:25.00073. Schedule next: false
979006:2024.10.29 13:22:25.077180 [ 8102 ] {0a761af5-2618-4751-823d-4903ce972706} <Trace> QueryMetricLog: Query 0a761af5-2618-4751-823d-4903ce972706 has a more recent metrics collected. Skipping this one


```

## A quick task that when the scheduled task ran the query had already finished

```bash
zstdgrep -ina "QueryMetricLog" clickhouse-server.log.zst | grep -i "a2270846-3fd5-4a2a-a226-3b1f41730d9e"                                                                           
1433540:2024.10.29 13:28:05.359437 [ 1124 ] {} <Debug> QueryMetricLog: Collecting query_metric_log for query a2270846-3fd5-4a2a-a226-3b1f41730d9e with QueryStatusInfo from 2024.10.29
 13:28:05.00359. Schedule next: true                                                                                                                                                  
1433544:2024.10.29 13:28:05.361650 [ 1124 ] {} <Trace> QueryMetricLog: Query a2270846-3fd5-4a2a-a226-3b1f41730d9e finished already while this collecting task was running  
```

```sql
SELECT
    event_time_microseconds,
    type,
    query_duration_ms,
    exception
FROM file('query_log.tsv.zst')
WHERE (query_id = 'a2270846-3fd5-4a2a-a226-3b1f41730d9e') AND (type > 1)
FORMAT Vertical

Query id: fa346149-3d1e-471f-a6da-65b8c3f826f3

Row 1:
──────
event_time_microseconds: 2024-10-29 13:28:05.335052
type:                    QueryFinish
query_duration_ms:       982
exception:               

1 row in set. Elapsed: 3.300 sec. Processed 290.70 thousand rows, 26.83 MB (88.09 thousand rows/s., 8.13 MB/s.)
Peak memory usage: 269.52 MiB.
```

antaljanosbenjamin reviewed (approved):
LGTM.
